### ISSUE TITLE
NoSuchElementException when trying to map an ScallopOption with a converter error and a co-dependent validation

### ISSUE THREAD

**Issue by evantill** [scallop/scallop#57]
In my case I try to define a Date opt using a custom converter
If the provided date is not well formatted I have this exception :

```
java.util.NoSuchElementException: Either.right.value on Left
    at scala.util.Either$RightProjection.get(Either.scala:454)
    at org.rogach.scallop.Scallop$$anonfun$isSupplied$4.apply(Scallop.scala:521)
    at org.rogach.scallop.Scallop$$anonfun$isSupplied$4.apply(Scallop.scala:519)
    at scala.Option.map(Option.scala:145)
    at org.rogach.scallop.Scallop.isSupplied(Scallop.scala:519)
    at org.rogach.scallop.Scallop$$anonfun$18.apply(Scallop.scala:381)
    at org.rogach.scallop.Scallop$$anonfun$18.apply(Scallop.scala:381)
    at scala.collection.TraversableLike$$anonfun$filter$1.apply(TraversableLike.scala:264)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$class.filter(TraversableLike.scala:263)
    at scala.collection.AbstractTraversable.filter(Traversable.scala:105)
    at org.rogach.scallop.Scallop.getAllSuppliedOptionNames(Scallop.scala:381)
    at org.rogach.scallop.Scallop$$anonfun$verify$13.apply(Scallop.scala:588)
    at org.rogach.scallop.Scallop$$anonfun$verify$13.apply(Scallop.scala:588)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
    at scala.collection.AbstractTraversable.map(Traversable.scala:105)
    at org.rogach.scallop.Scallop.verify(Scallop.scala:587)
    at org.rogach.scallop.ScallopConf.verify(ScallopConf.scala:274)
    at org.rogach.scallop.ScallopConf.afterInit(ScallopConf.scala:445)
    at org.rogach.scallop.AfterInit$class.delayedInit(AfterInit.scala:14)
    at org.rogach.scallop.ScallopConf.delayedInit(ScallopConf.scala:26)
```

I would expect a WrongOptionFormat exception

**Comment by Rogach** [comment]
Thanks for reporting the issue! I reproduced it, and will fix it in approx. 12 hours (can't do it right now)

**Comment by evantill** [comment]
:+1: 

no problems I am using a workaround

**Comment by Rogach** [comment]
Could you please post the code you used to get this exception? It would help a lot.

**Comment by evantill** [comment]
strange, I have forked and created a test from the master branch and the problem is not reproducible.
Have you worked on it ?

**Comment by evantill** [comment]
I still have a test in my application with the error but could not reproduce it for the moment in a unit test outside my code.

The exception is raised at line [Scallop.scala#L523](https://github.com/Rogach/scallop/blob/master/src/main/scala/org.rogach.scallop/Scallop.scala#L523)

**Comment by evantill** [comment]
I will change the title of this issue. The problem is related to the validation of codependent options

**Comment by evantill** [comment]
I think the problem could be related to the template ScallopConfValidations.fm
the validate functions check only the first argument before calling fn()

IMHO ScallopOptions should extends  Function0[Option[A]] and not  Function0[A]
in this case the validate function could be redefined in a safer manner

``` scala
new Function0[Either[String, Unit]] {
  def apply = ( for( v1 <- o1; v2 <- o2 ) yield fn(v1,v2) ).getOrElse(Right(Unit) )
```

**Comment by evantill** [comment]
is it possible to consider heavy changes like this ? If so I could contribute.

**Comment by Rogach** [comment]
No, in fact, the problem was that for some reason the option group validation ran before option format checking (in `Scallop.verify`). I fixed that, and now the test works as expected. I'll be pushing the branch now, and publishing a snapshot.

About `ScallopOption` and functions - personally, I feel that it's an _extremely_ bad idea to break all of the existing code (by swapping the meaning of `apply` and `get`). I do understand that current meaning of those methods contradicts the meaning of corresponding methods on `scala.Option`, but this was done for a reason - most of the time, options are either defined as required, or have explicitly specified default value, thus `.get` is used much less often. And the user can treat the whole ScallopConf as his config object, almost without any boilerplate in the code.

So, currently I'm against making ScallopOption subclass functions - mostly because of existing code, though.

**Comment by Rogach** [comment]
I published a snapshot with a fix (for scala 2.10.x):

```
resolvers += "sonatype snapshots" at "https://oss.sonatype.org/content/repositories/snapshots"
libraryDependencies += "org.rogach" %% "scallop" % "0.9.1~1~SNAPSHOT"
```

**Comment by evantill** [comment]
thank you I will give it a try
